plugins/osdn/ovs/proxy.go

```diff
@@ -0,0 +1,151 @@
+)
+
+type ovsProxyPlugin struct {
+	registry *osdn.Registry
+	podsByIP map[string]*kapi.Pod
+	mutex    sync.Mutex
+
+	baseEndpointsHandler pconfig.EndpointsConfigHandler
+}
+
+func CreateProxyPlugin(registry *osdn.Registry) (api.FilteringEndpointsConfigHandler, error) {
+	return &ovsProxyPlugin{
+		registry: registry,
+		podsByIP: make(map[string]*kapi.Pod),
+	}, nil
+}
+
+func (proxy *ovsProxyPlugin) Start() error {
+	glog.Infof("Starting multitenant SDN proxy endpoint filter")
+
+	// Populate pod info map synchronously so that kube proxy can filter endpoints to support isolation
+	pods, err := proxy.registry.GetAllPods()
+	if err != nil {
+		return err
+	}
+
+	for _, pod := range pods {
+		proxy.trackPod(&pod)
+	}
+
+	go proxy.watchPods()
+
+	return nil
+}
+
+func (proxy *ovsProxyPlugin) watchPods() {
+	podevent := make(chan *osdn.PodEvent)
+	go proxy.registry.WatchPods(podevent)
+	for {
+		ev := <-podevent
+		switch ev.Type {
+		case osdn.Added:
+			proxy.trackPod(ev.Pod)
+		case osdn.Deleted:
+			proxy.unTrackPod(ev.Pod)
+		}
+	}
+}
+
+func (proxy *ovsProxyPlugin) getTrackedPod(ip string) (*kapi.Pod, bool) {
+	proxy.mutex.Lock()
+	defer proxy.mutex.Unlock()
+
+	pod, ok := proxy.podsByIP[ip]
+	return pod, ok
+}
+
+func (proxy *ovsProxyPlugin) trackPod(pod *kapi.Pod) {
+	if pod.Status.PodIP == "" {
+		return
+	}
+
+	proxy.mutex.Lock()
+	defer proxy.mutex.Unlock()
+	podInfo, ok := proxy.podsByIP[pod.Status.PodIP]
+
+	if pod.Status.Phase == kapi.PodPending || pod.Status.Phase == kapi.PodRunning {
+		// When a pod hits one of the states where the IP is in use then
+		// we need to add it to our IP -> namespace tracker.  There _should_ be no
+		// other entries for the IP if we caught all of the right messages, so warn
+		// if we see one, but clobber it anyway since the IPAM
+		// should ensure that each IP is uniquely assigned to a pod (when running)
+		if ok && podInfo.UID != pod.UID {
+			glog.Warningf("IP '%s' was marked as used by namespace '%s' (pod '%s')... updating to namespace '%s' (pod '%s')",
+				pod.Status.PodIP, podInfo.Namespace, podInfo.UID, pod.ObjectMeta.Namespace, pod.UID)
+		}
+
+		proxy.podsByIP[pod.Status.PodIP] = pod
+	} else if ok && podInfo.UID == pod.UID {
+		// If the UIDs match, then this pod is moving to a state that indicates it is not running
+		// so we need to remove it from the cache
+		delete(proxy.podsByIP, pod.Status.PodIP)
+	}
+}
+
+func (proxy *ovsProxyPlugin) unTrackPod(pod *kapi.Pod) {
+	proxy.mutex.Lock()
+	defer proxy.mutex.Unlock()
+
+	// Only delete if the pod ID is the one we are tracking (in case there is a failed or complete
+	// pod lying around that gets deleted while there is a running pod with the same IP)
+	if podInfo, ok := proxy.podsByIP[pod.Status.PodIP]; ok && podInfo.UID == pod.UID {
+		delete(proxy.podsByIP, pod.Status.PodIP)
+	}
+}
+
+// FilteringEndpointsConfigHandler implementation
+func (proxy *ovsProxyPlugin) SetBaseEndpointsHandler(base pconfig.EndpointsConfigHandler) {
```

Comment:
Can we pass `base pconfig.EndpointsConfigHandler` to CreateProxyPlugin and get rid of this method?